pkg/sentry/platform/kvm/machine.go

```diff
@@ -296,13 +296,17 @@ func newMachine(vm int) (*machine, error) {
 	return m, nil
 }
 
-// hasSlot returns true iff the given address is mapped.
+// hasSlot returns true if the given address is mapped.
 //
 // This must be done via a linear scan.
 //
 //go:nosplit
 func (m *machine) hasSlot(physical uintptr) bool {
-	for i := 0; i < len(m.usedSlots); i++ {
+	slotLen := int(atomic.LoadUint32(&m.nextSlot))
```

Comment:
m.nextSlot is used like a lock:
https://cs.opensource.google/gvisor/gvisor/+/master:pkg/sentry/platform/kvm/bluepill_fault.go;l=94;drc=9ba8c40a3a3c7fed40d9137fed8a87fa9d536a22;bpv=1;bpt=1

Reply:
I don't object to this change, but the improvement is incremental at best. We probably won't change much by not scanning all entries? Ultimately it's a very small array.

Note that the locking semantics here are optimistic. If hasSlot returns false, it will attempt to map again, which will acquire the slot lock and catch the conflict (EEXIST).

Reply:
Sorry, I meant this specific improvement. I suspect most of the improvement is coming from storing the correct address, which is a valuable and excellent fix (thanks for doing this!). I just wanted to be clear. I'll let @zhlhahaha decide what to do about this specific point.

Reply:
I don't object either, but I think it would be good to have a comment here to explain when slotLen can be greater than maxSlots and why it isn't an issue here.

Reply:
>  I suspect most of the overall improvement is coming from storing the correct address

Yes, Adin, I did benchmark for this specific tunning, the performance improvement is not obvious (actually I can not see any difference with or without this tunning). I just think it is not good to scan the whole slice even though most parts of the slice are not allocated.

> I don't object either, but I think it would be good to have a comment here to explain when slotLen can be greater than maxSlots and why it isn't an issue here.

Thanks, Andrei, I will add comments here.